03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,382 @@
+
+```{r}
+# We can use the `summary()` function to get a better idea of the distribution of values in the `Entrez` column
+summary(as.factor(mapped_df$Entrez)) # We need to use the `as.factor()` function here in order to get the count of each unique value returned
+```
+
+There are 9102 NA's in our data frame, which means that 9102 out of the 31882 Ensembl IDs did not map to Entrez IDs.
+This means if you are depending on `Entrez` IDs for your downstream analyses, you may not be able to use the 9102 unmapped genes.
+
+Now let's check to see how many genes we have that were mapped to multiple IDs.
+
+```{r}
+multi_mapped_df <- mapped_df %>%
+  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
+  dplyr::count(Ensembl, name = "entrez_id_count") %>%
+  # Arrange by the genes with the highest number of Entrez IDs mapped
+  dplyr::arrange(desc(entrez_id_count))
+
+# Let's look at the first 6 rows of our `multi_mapped` object
+head(multi_mapped_df)
+```
+
+Looks like we have one case where one Ensembl ID mapped to 15 Entrez IDs!
+We have a total of 235 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
+With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.
+If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+### Store all mapped values
+
+If we are not sure which Entrez ID is most relevant to our downstream analyses, we could store all of the mapped information for future use.
+In the next chunk, we show how we can collapse all the Entrez IDs into one column where we store them all for each unique Ensembl ID.
+
+```{r}
+collapsed_mapped_df <- mapped_df %>%
+  # Group by Ensembl IDs
+  dplyr::group_by(Ensembl) %>%
+  # Collapse the mapped values in `mapped_df` into one column named `all_entrez_ids`
+  dplyr::summarize(all_entrez_ids = paste(Entrez, collapse = ";"))
+```
+
+Let's take a look at our new collapsed `all_entrez_ids` column in the `collapsed_mapped_df` object.
+
+```{r}
+collapsed_mapped_df %>%
+  # Filter `collapsed_mapped_df` to include only the rows where `all_entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
+  dplyr::filter(stringr::str_detect(all_entrez_ids, ";")) %>%
+  # We only need a preview here
+  head()
+```
+
+You may have a list of Entrez IDs you are interested in, in which case, the above format may work best for you.
+In a different study, you may want the oldest Entrez ID (which is probably first), in which case, you can create a column that stores just the first mapped Entrez ID that comes up for each Ensembl ID.
```

Comment:
Do we have confirmation that the first Entrez ID is the lowest number? i.e. is the list we get here sorted?

Reply:
Good question! It does not explicitly that the returned list is sorted in the `AnnotatationDbi` or `mapIDs()` docs but the lists of mapped values returned does appear to be sorted from my investigation.